File: src/main/kotlin/com/ahold/technl/sandbox/validation/DeliveryStatusValidator.java
package com.ahold.technl.sandbox.validation;

import jakarta.validation.ConstraintValidator;
import jakarta.validation.ConstraintValidatorContext;

import java.util.Arrays;

public class DeliveryStatusValidator implements ConstraintValidator<ValidDeliveryStatus, Enum<?>> {
    private Class<? extends Enum<?>> enumClass;

    @Override
    public void initialize(ValidDeliveryStatus annotation) {
        this.enumClass = annotation.enumClass();
    }

    @Override
    public boolean isValid(Enum<?> value, ConstraintValidatorContext context) {
        if (value == null) return true;
        return Arrays.stream(enumClass.getEnumConstants())
                .anyMatch(e -> e.name().equals(value.name()));
    }
}

File: src/main/kotlin/com/ahold/technl/sandbox/validator/DeliveryStatusValidator.kt
package com.ahold.technl.sandbox.validator

import com.ahold.technl.sandbox.dto.DeliveryStatus
import jakarta.validation.ConstraintValidator
import jakarta.validation.ConstraintValidatorContext

class DeliveryStatusValidator : ConstraintValidator<ValidDeliveryStatus, DeliveryStatus> {

    override fun isValid(status: DeliveryStatus?, context: ConstraintValidatorContext): Boolean {
        return status == DeliveryStatus.IN_PROGRESS || status == DeliveryStatus.DELIVERED
    }
}
